### ISSUE TITLE
Architecture direction (discussion)

### ISSUE THREAD

**Issue by ryardley** [blitz-js/blitz#306]
### What do you want and why?

This is a big one but could work as a roadmap for what we want in terms of the internal codebase of blitz moving forward.

We need to guide our internal architecture to avoid unnecessary dependencies and spaghetti code.

We want to somehow keep things orthogonal which means following "things that change together stay together" but also allow for horizontal layers to be grouped where appropriate.

### Possible implementation(s)

I have a suggestion (feel free to workshop this with me - also if you are a code owner and want to update the description of any package please comment):

# ~~app~~ framework

~~User app~~ Code that actually runs in user apps.  
* rpc, 
* cache management
* rpc and reactQuery templates used by generator and server-manager's synchronizer rules should live here and be imported

Code should be built for both node and browser and deeply linked. 

# blitz 
__Blitz wrapper package.__

Determines launch environment exports public api for users. This stays the same

# cli 
__Argument parsing__

Responsible for parsing command-line arguments and delegating data to the correct downstream packages and displaying help text. Data qualification with enquirer can be managed here as that is part of parsing arguments (or not?) 

Will probably accept configuration from the prisma package and generate an oclif plugin internally here to configure new commands.

# repl

__REPL__

Launch the blitz REPL 

# generator

__Dynamically generate source files in a users project__

Generate files within client app directory. Pull in templates and generators from the prisma package and elsewhere eventually and use that to generate assets appropriately. 

# prisma
__Database plugin__

Fledgeling db package and first plugin. Anything related to prisma should live here and get exported into wherever it is needed. API can be refactored internally and get ready for release

We want to avoid any system dependencies here that we don't control (eg. oclif) as whatever dependencies we have here that aren't prisma related will need to eventually become part of the plugin framework. Naturally anything prisma/db related is fine

# synchronizer 

__Accept rules and pipe and transform files based on those rules__

Accept config and setup a vinyl file transform pipeline with dynamic stages (could possibly eventually be used by generator) pipe them to a file writer interface. 

# server-manager

__Manage next-server process__

Accept an input stream listen for events and manage the next server start, stop and output appropriately

# ~~utils~~

~~__Shared general development libraries__~~

~~General small focussed development utilities similar to lodash and not blitz specific. Nothing should have any dependencies on anything else in the codebase aside from potentially other utilities.~~

EDIT: Avoiding a utils library will mean less refactoring in the future as we rarely will need to extract out libs from utils which may happen if people start adding things to utils.

# display

__Render client facing awesomeness__

Currently called `log`. Blitz specific node cli display library. This might end up depending on ink or something like that at a later date. Should be able to be piped to. Can be swapped out for a view that writes JSON objects to a file for debugging for example. Can be set to remove all clear commands for nice display logging in a concurrent process or logging to a file.

# UPDATE: Possible Action Summary

- [ ] Create `app` package - deep imports separating context - move `rpcHandler` from server.

**Comment by aem** [comment]
I've got a branch locally that splits the `generator` package out of the CLI. I'll wait to open the PR until #305 and #320 get in so we can cleanly move over the generators without affecting open code, but it's a pretty clean break after that. One thing we should note is that we've already got some utils in the CLI that are used both by the generator and the other CLI commands. For now I opted to just duplicate them, as they aren't too complex, but I expect at some point we'll need a `@blitzjs/node-utils` package that we can use to dump common utilities that we need to share across the generator, CLI, prisma, and REPL packages.

I'm generally against all-purpose `utils` packages since they tend to get bloated and overcrowded and would rather a large number of highly specific packages. While I think this example is a bit too far in the other direction, the [lerna project](https://github.com/lerna/lerna/tree/master/utils) runs similarly to this. I'd probably argue for a _little_ more colocation than that 🙂, maybe grouping the utils by concern (e.g. `(node|cli)-utils` like I suggested above) 

Having said that, it's a preference but not something I really care too deeply about so if people would rather have just a catch-all utils package then that's ok.

**Comment by ryardley** [comment]
> I'm generally against all-purpose utils packages since they tend to get bloated and overcrowded and would rather a large number of highly specific packages. 

I have to say I really agree with this sentiment. I think it might lead to easier maintenance. 

> While I think this example is a bit too far in the other direction, the lerna project runs similarly to this.

I am not sure I think it is not that bad honestly feels more future proof means PRs will be localized to single packages. 

As far as the idea of of separating execution context is concerned I think we are going to continue running into the issue of wanting certain things colocated but requiring different execution contexts (the above `app` package comes to mind ) and I think the solution is that we need to have to have a standard way to manage deep imports. TSDX really holds us back here. Can we get away with just using tsc to manage this? Do we even want to bundle our code at this sub-package level? Some research here might help us out a fair bit.

**Comment by aem** [comment]
> Can we get away with just using tsc to manage this? Do we even want to bundle our code at this sub-package level? Some research here might help us out a fair bit.

Not sure what the community convention is but I'm 100% on board for skipping the whole build + tooling pipeline for the micropackages. `tsc` would probably do just fine

**Comment by ryardley** [comment]
> Not sure what the community convention is but I'm 100% on board for skipping the whole build + tooling pipeline for the micropackages. tsc would probably do just fine

I agree

**Comment by flybayer** [comment]
Thanks for starting this Rudi!

Having a many folders in the repo is nice because we can more easily delegate code ownership.

The only thing I really care about is that we have a bulletproof way to keep server code out of client-side app bundles. I.e. it should be impossible for anyone to do this on accident.

I like the idea of not having to publish 100 npm packages.

**Comment by ryardley** [comment]
Ok so could one action point be to setup a simple tsc dev/build/test process for a single package refactoring? Currently the "server" component of core is located in "server" perhaps we can create an issue to bring that together with core so we have a model of colocating different execution contexts?

**Comment by ryardley** [comment]
> I like the idea of not having to publish 100 npm packages.

Could the friction of having to create a package help us avoid that?

**Comment by ryardley** [comment]
At risk of changing the topic a little - What do folks esp @merelinguist think about the idea of creating a global reactive frontend that we stream events to? 

```tsx
someBackendProcess.pipe(display.stream)
```

At a later point we might want to refactor to use Ink

```tsx
render(<Display events={inputStream} />)
```

This means we can swapout the display engine for an event logger or stream to file or anything really

```tsx
someBackendProcess.pipe(DEBUG ? eventLogger.stream : display.stream)
```

**Comment by flybayer** [comment]
> Ok so could one action point be to setup a simple tsc/jest dev/build/test process with deep imports for a single package refactoring? Currently the "node" component of `core` (`rpcHander` I think it's called) is located in `@blitzjs/server` but it really will change with the rpc client in core - perhaps we can create an issue to bring that together with core in `app`(if no one objects to the name) so we have a model of colocating different execution contexts?

Yeah good for someone to work on this.

**Question**: why rename `core` to `app`?

> At risk of changing the topic a little - What do folks esp @merelinguist think about the idea of creating a global reactive frontend that we stream events to?

For development only?

**Comment by merelinguist** [comment]
> At risk of changing the topic a little - What do folks esp @merelinguist think about the idea of creating a global reactive frontend that we stream events to?

@ryardley I’ve actually been thinking a *lot* about this, especially while building the GUI, and I think that’s a golden idea!

**Comment by ryardley** [comment]
> Question: why rename core to app?

In my head it is because any code that changes alongside code that actually runs in an __app__ should live there (including generator/server-manager templates). Perhaps the semantics aren't right I think @merelinguist could have something good to say about that. Maybe `framework` also make sense. I think `core` could end up a catchall and is not that descriptive. Not attached at all happy to roll with what what folks think. 

> global reactive frontend that we stream events to

I guess that is for development in that it would not be running in production as next takes care of production output.

Basically instead of just using a `log` library we design our frontend as a reactive app. This will make refactoring to something like `Ink` easier in the longrun.

**Comment by ryardley** [comment]
> @ryardley I’ve actually been thinking a lot about this, especially while building the GUI, and I think that’s a golden idea!

I think having some kind of simple state machine in the view will help structure too

**Comment by ryardley** [comment]
> @ryardley I’ve actually been thinking a lot about this, especially while building the GUI, and I think that’s a golden idea!

The tradeoff to be clear is that we get a certain coupling around events as they are the value objects of the system

**Comment by flybayer** [comment]
I like `framework`. My concern with `app` is "app" usually is describing an end-user app. This could be confusing to see in our code for beginners and passer-bys. But `core` or `framework` both convey that this is custom blitz internals

**Comment by ryardley** [comment]
> I like framework. My concern with app is "app" usually is describing an end-user app. This could be confusing to see in our code for beginners and passer-bys. But core or framework both convey that this is custom blitz internals

I agree

**Comment by ryardley** [comment]
Ok so for me this one is a toughy:

I think enquirer should be wrapped by our display application and we should be using an enquirer like interface on that to write out questions however...

Do we ask the questions within the cli or do downstream systems manage that process of qualifying input? 

My gut says cli should do it but I am close to on the fence with this one. It could be both.

**Comment by aem** [comment]
Mabye a middle ground where we have a `@blitzjs/prompt` package that's a thin wrapper around `enquirer` and provides some stock prompts (generic confirm prompt, select a file path, etc, just a few that we'll need regularly), and then anything as far as the actual input qualification and validation we leave up to the consuming package?

**Comment by MrLeebo** [comment]
@flybayer @ryardley I'm closing this because I think the architectural goals that have been outlined here have been achieved. If you disagree, could you please reopen with your assessment of where we're at and what still needs to be done?
